### PR TITLE
Fix #41: output order and subset

### LINKED ISSUE #41: On output order
In unparse()

```
    // create heading
    if ($this->heading && !$append && !empty($fields)) {
        foreach ($fields as $key => $value) {
            $entry[] = $this->_enclose_value($value, $delimiter);
        }

        $string .= implode($delimiter, $entry).$this->linefeed;
        $entry   = array();
    }
    // create data
    foreach ($data as $key => $row) {
        foreach ($row as $field => $value) {
            $entry[] = $this->_enclose_value($value, $delimiter);
        }

        $string .= implode($delimiter, $entry).$this->linefeed;
        $entry   = array();
    }
```

if there is $fields exists, it seems the output should based on the $fields, otherwise
1. How do we guarantee the data row order is the same to the $fields
2. If $fields is given to only select some fields not all, this unparse seems broken

Can someone first check if above is an issue? i could provide some kind of fix.

### PR DESCRIPTION
**Pull request by susgo**
~~This is currently **work in progress**.~~

At the moment this request only implements tests of features which should be done for solving the comments on #41. In my opinion the following things should be work:
* [x] If no `fields` are given take the `title` property. Works.
* [x] If the given `fields` define same number of strings like `titles` attribute, but the names are different (the strings differ) from the `titles` property then the fields should be renamed. Works as default.
* [x] If the given `fields` are an subset of the `titles` property, only subset should be unparsed.
* [x] If the given `fields` are a reorder of the `titles` property, data should be reordered, too, on unparsing
* [x] Allow renaming and reordering by using array keys in `fields`

This request is ready, when all tests are green. 

Please feel free to close this pull request if the feature is not needed anymore or I miss something.

### CODE REVIEW

File: src/Csv.php [hunk 1]
```diff
@@ -777,8 +777,15 @@ public function unparse($data = array(), $fields = array(), $append = false, $is
         $entry = array();
 
         // create heading
+        $fieldOrder = $this->_validate_fields_for_unparse($fields);
+        if (!$fieldOrder && !empty($data)){
```
Comment (Fonata) [inline src/Csv.php]: Please add a space before the curly bracket.

**Comment by susgo** [comment]
This is ready. I missed to write it here.

File: src/Csv.php [hunk 2]
```diff
@@ -777,8 +777,15 @@ public function unparse($data = array(), $fields = array(), $append = false, $is
...
+            $column_count = count($data[0]) - 1;
```
Comment (Fonata) [inline src/Csv.php]: Judging by the variable name, the `- 1` should not be on this line, but within the `range()` call.

**Comment by susgo** [comment]
Corrected now

**Comment by Fonata** [comment]
@jimeh @williamknauss I'll merge this in March 10 at the earliest unless you share your thoughts here.

@susgo These Would you be ok with me squashing your commits then? The output of `git log --oneline --graph` looks rather complicated due to the branches and merges:
````
* e29a33d (HEAD -> 41-output-order-and-subset, itexia/41-output-order-and-subset) small code improvements after review
* 9e9c188 implements functionality of this issue
* bdbffa9 added one test for heading=false
* 732fb19 init unparse tests for ordering and subseting by fields
*   ef44ea3 Merge pull request #7 from parsecsv/master
|\
| * 2ac9f45 Check if output contains HTML table cells
| * 3968659 Make sure the code examples run without syntax or runtime errors.
| * d3a084d Added namespace support for further tests
| * b0355be Re-added the parsecsv.lib.php file for compatibility
| * b753d9c Repaired examples: there was a require missing.
| *   2b95ec5 Merge remote-tracking branch 'origin/Tests-for-sort-and-cond' into pull-test
| |\
|/ /
| * 913dc94 Keep sorting keys compatible: Moved array_values call to PHPUnit.
| * 5ced1f4 Added conditions to SortByTest because otherwise different PHP versions...
| * ca84469 Moved ConditionsTest to properties subfolder
| * 75902f4 Added PHPUnit test for sorting; I could not get the rows in the right...
| * 7470d2b PHPUnit: Repaired file paths in autoQuotesDataProvider()
| * 01f0891 Test some conditions
* | 3f595df only renamed variable
* |   0819493 Merge pull request #6 from parsecsv/itexia-fixed-remaining-test
|\ \
| * | a60aae4 getDatatypes: Documented the need for PHP 5.5 or higher
| * | 4cba97a Allow PhpStorm to understand the code better
| * | 2206ec1 only reformatted source code to match the rest of the project
| * | d188921 composer.json: Only allow to load tests in dev environment
| * | 85cc0d9 Updated README.md to new name & namespace
| * | a7f2bbb Throw more specific exception
| * | d484f3e The workaround for array_count_values was wrong and unnecessary.
| * | f8b2fe2 Added missing letter in function name
|/ /
* |   17bfd94 Merge pull request #5 from parsecsv/itexia-repair-tests
|\ \
| * | ec7b1c1 PHPUnit: Better class names with modern naming scheme
| * | ba9eae1 Bootstrap: Improved the conditions under which the test wrapper loads
| * | 8f26ce6 Force Travis to output which PHPUnit version it uses
|/ /
* | 52a9b61 added DatatypeEnum tests; repaired the other test for local execution; small bug fixes
* |   5660be3 Merge pull request #1 from itexia/introduce-namespaces
|\ \
| * \   70b6fe4 (itexia/introduce-namespaces) Merge branch 'introduce-namespaces' of https://github.com/itexia/parsecsv-for-php into introduce-namespaces
| |\ \
| | * \   31dba04 Merge pull request #4 from parsecsv/test-helper-for-itexia-with-namespaces
| | |\ \
| | | * \   6da0537 Merge branch 'introduce-namespaces' into test-helper-for-itexia-with-namespaces
| | | |\ \
| | | |/ /
| | |/| |
| | | * |   2b6e387 Merge remote-tracking branch 'origin/test-helper-for-itexia-master' into delete_this_branch
| | | |\ \
| | | | * | d4f9c9f Updated worthless_test to allow for more fields
| | | | * | 2b17f01 Removed additional column to make expected data fit to fixture
| | | | * | 9551862 Allow current working dir to be different when running tests
| | | | * | 736bc48 getDatatypes: throw more meaningful exception if no data is present
| | | | * | f702ca9 Float regex: Don't allow double dot; do allow 'e' for exponentials
| |_|_|/ /
|/| | | |
| | | * | af73683 Changed tests to work now that we have a namespace
| | | * | f2fb7c8 Absolute path in require do stop PhpStorm from nagging
| | | * | 92653c9 Force predictable working dir; create autoloader file if it doesn't exist
| * | | | f1a8912 regex updated according to merge conflict
| |/ / /
| * | |   91091ac Merge pull request #2 from itexia/new-enom-datatype
| |\ \ \
| | |/ /
| |/| |
| | * | 2c7c552 Update DatatypeTrait.php
| | * | 52e8795 corrected datatype count
| | * | 48cec20 namespace change
| | * |   c266ef9 Merge branch 'introduce-namespaces' of https://github.com/itexia/parsecsv-for-php into new-enom-datatype
| | |\ \
| | |/ /
| |/| |
| * | | 76a0406 updated examples for using new classname
| * | | 17ac1de renamed namespace
| | * | f6bd414 init enum for data type
````

As a remedy, please consider a rebase-centered strategy in your repo to linearize things.

**Comment by susgo** [comment]
@Fonata okay. Sorry for the extra effort. Should concentrate on one thing only.

**Comment by Fonata** [comment]
No worries, your input is highly appreciated here.

**Comment by susgo** [comment]
What's with this pull request? Sorry, I can do some refactoring work or do some code in traits but firstly all should be merged.

**Comment by Fonata** [comment]
Thanks for the reminder. All is well with this PR. [Above](https://github.com/parsecsv/parsecsv-for-php/pull/126#issuecomment-370141484) I wrote to merge this on March 10 at the earliest. I will go ahead and merge it now.

**Comment by Fonata** [comment]
Ups, now the [Travis test for the master branch failed](https://travis-ci.org/parsecsv/parsecsv-for-php) because `testSaveWithNewHeader` found your code changes things. If you don't have time to investigate, please leave a note here and I'll attempt fixing it.

**Pull request merged**